### PR TITLE
[MM-36457] - Send email when license is up for renewal

### PR DESCRIPTION
**Pull request by AGMETEOR**
#### Summary
Send an email to admins when license up for renewal (60 days before expiration)

Screenshot 2021-06-17 at 13 59 16

#### Ticket Link
[MM-36457](https://mattermost.atlassian.net/browse/MM-36457?atlOrigin=eyJpIjoiY2Y5ZDE3N2UyMDQ3NGM5ODgyNDNiZDNlMjQ0NTFhYzciLCJwIjoiaiJ9)

#### Release Note
```release-note
NONE

```

### CODE REVIEW

**Comment by marianunez** [comment]
@AGMETEOR can we add a screenshot here of the email for UX Review by @annelieseklein ?

File: app/server.go
Comment (nickmisasi) [inline app/server.go:1828]: Genericize this function name a bit more. 

`IsWithinExpirationPeriod` or something along those lines.

File: model/license.go
Comment (nickmisasi) [inline model/license.go:273]: Let's const these values, and you'll have to adjust the function name for IsAroundSixtyDaysToExpiration.

File: i18n/en.json
Comment (nickmisasi) [inline i18n/en.json:3335]:
```suggestion
    "translation": "{{.UserName}}, your subscription is set to expire in {{ some variable amount }} days. We hope you’re experiencing the flexible, secure team collaboration that Mattermost enables. Renew soon to ensure your team can keep enjoying these benefits."
```

Probably shouldn't hard code the number of days

**Review by nickmisasi — CHANGES_REQUESTED** [review]
Further question:

From what I can tell, `doLicenseExpirationCheck` is run every 24 hours from server start. Won't this cause the admins to get an email every day?

File: app/server.go
Comment (marianunez) [inline app/server.go:1787]: Let's wrap this error with more context on where we are here generating the renewal link

File: app/server.go
Comment (marianunez) [inline app/server.go:1829]: As discussed offline, given that this is executes in the server runtime level and in an HA environment we can end up with multiple emails, could we store a flag in database indicating if this has been sent (or a timestamp of when it was)? We do something like this work: https://github.com/mattermost/enterprise/blob/cf4adb760d3e58eab50d71428ab80da805296fd6/cloud/worker.go#L96

File: i18n/en.json
Comment (marianunez) [inline i18n/en.json:16]: Is this intended?

File: templates/license_up_for_renewal.mjml
Comment (marianunez) [inline templates/license_up_for_renewal.mjml:61]: nit: missing new line

File: app/server.go
Comment (AGMETEOR) [reply app/server.go:1829]: Alright. Going to do that

File: i18n/en.json
Comment (AGMETEOR) [reply i18n/en.json:16]: No. Going to update

File: app/server.go
Comment (AGMETEOR) [reply app/server.go:1787]: Can't really wrap it because it's already `*model.AppError`

**Review by nickmisasi — APPROVED** [review]

File: app/server.go
Comment (marianunez) [reply app/server.go:1794]: You could create a new app error and include the error of the original one, 
```suggestion
		return model.NewAppError("s.sendLicenseUpForRenewalEmail", "Failed to generate the license renewal link", nil, appErr.Error(), http.StatusInternalServerError)
```

File: app/server.go
Comment (marianunez) [inline app/server.go:1788]: Should we handle the error case here? 

Also, can we make a comment on what is happening here? We are returning early because the email was already sent

File: app/server.go
Comment (marianunez) [inline app/server.go:1823]: Why a `Critical` level here?

**Comment by marianunez** [comment]
@AGMETEOR can you help update the QA Steps in the ticket for Steve ?

File: app/server.go
Comment (AGMETEOR) [reply app/server.go:1788]: I think for this we want the case where the err is nil because it confirms that we already did this action and we therefore don't want to proceed any further. The opposite where the err is not nil we want to continue because the key cannot be retrieved.

Add a comment to the code as suggested.

File: app/server.go
Comment (AGMETEOR) [reply app/server.go:1823]: Changed to Debug level

**Comment by AGMETEOR** [comment]
> @AGMETEOR can you help update the QA Steps in the ticket for Steve ?

@stevemudie Added the QA steps. Let me know in case they're not clear

File: app/server.go
Comment (marianunez) [inline app/server.go:1848]: Let's provide context in this error here as well

```suggestion
			mlog.Debug("Error sending renewal email", appErr.Error())
```

**Review by marianunez — APPROVED** [review]
LGTM. Minor suggestions below.

**Comment by marianunez** [comment]
Setting the `Do not merge` so we can revert the license key change before merging. 

@AGMETEOR can you take a look at the tests?

File: app/server.go
Comment (marianunez) [reply app/server.go]: Thanks for the explanation here. Sounds good 👍

**Comment by AGMETEOR** [comment]
> Setting the `Do not merge` so we can revert the license key change before merging.
> 
> @AGMETEOR can you take a look at the failing tests?

@marianunez tests are failing due to license trial license being generated not being valid. `"caller":"utils/license.go:99","msg":"Invalid signature"` this occurs at the point of saving the license

**Comment by AGMETEOR** [comment]
Hello @stevemudie I can see steve.mudie+dan@mattermost.com and steve.mudie+shane@mattermost.com and license is at 59 days to expiration. Did you receive any emails?

**Comment by stevemudie** [comment]
> Hello @stevemudie I can see [steve.mudie+dan@mattermost.com](mailto:steve.mudie+dan@mattermost.com) and [steve.mudie+shane@mattermost.com](mailto:steve.mudie+shane@mattermost.com) and license is at 59 days to expiration. Did you receive any emails?

@AGMETEOR 
I did (just checked now and they landed last night) but the link to renew was unable to load the page to renew. Is that expected?

**Comment by AGMETEOR** [comment]
> > Hello @stevemudie I can see [steve.mudie+dan@mattermost.com](mailto:steve.mudie+dan@mattermost.com) and [steve.mudie+shane@mattermost.com](mailto:steve.mudie+shane@mattermost.com) and license is at 59 days to expiration. Did you receive any emails?
> 
> @AGMETEOR
> I did (just checked now and they landed last night) but the link to renew was unable to load the page to renew. Is that expected?

Is it possible to get a screenshot of the emails to see if the styling is proper. Also, what error pops up?

**Comment by stevemudie** [comment]
Here's what I'm seeing.

I am logged into the server but the link still arrives at an error screen.

Also, of the two emails one of them looks proper while the other one is partially collapsed which is related to https://mattermost.atlassian.net/browse/MM-35535 but not a real concern here.

If it helps any here's the link that leads to the error: https://customers.mattermost.com/subscribe/renew?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJsaWNlbnNlX2lkIjoiZTVnY2N3NzFxYnJmZHE3c25xMW05ZXc1M28iLCJhY3RpdmVfdXNlcnMiOjQsImV4cCI6MTYyNTY5NTgzMH0.X7mprvTWK0EYrU_29JbhP5fTaZJ2jdf1PwXYxvc8SgU

Renew error
Up for renewal email

**Comment by marianunez** [comment]
> If it helps any here's the link that leads to the error:

The error is expected given that the link is pointing to the production Customer Portal which does not have your test licenses.

**Comment by AGMETEOR** [comment]
> > If it helps any here's the link that leads to the error:
> 
> The error is expected given that the link is pointing to the production Customer Portal which does not have your test licenses.

Oh yes. Thanks Maria

**Comment by AGMETEOR** [comment]
@stevemudie let’s give it one more day to make sure those emails don’t come in again and then that will be the final step of the QA

**Comment by AGMETEOR** [comment]
> Here's what I'm seeing.
> 
> I am logged into the server but the link still arrives at an error screen.
> 
> Also, of the two emails one of them looks proper while the other one is partially collapsed which is related to https://mattermost.atlassian.net/browse/MM-35535 but not a real concern here.
> 
> If it helps any here's the link that leads to the error: https://customers.mattermost.com/subscribe/renew?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJsaWNlbnNlX2lkIjoiZTVnY2N3NzFxYnJmZHE3c25xMW05ZXc1M28iLCJhY3RpdmVfdXNlcnMiOjQsImV4cCI6MTYyNTY5NTgzMH0.X7mprvTWK0EYrU_29JbhP5fTaZJ2jdf1PwXYxvc8SgU
> 
> Renew error
> Up for renewal email

I see the feedback email is missing in the email footer. @marianunez is that due to the fact that it’s blank in the config?

**Comment by marianunez** [comment]
> > Here's what I'm seeing.
> > I am logged into the server but the link still arrives at an error screen.
> > Also, of the two emails one of them looks proper while the other one is partially collapsed which is related to [mattermost.atlassian.net/browse/MM-35535](https://mattermost.atlassian.net/browse/MM-35535) but not a real concern here.
> > If it helps any here's the link that leads to the error: [customers.mattermost.com/subscribe/renew?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJsaWNlbnNlX2lkIjoiZTVnY2N3NzFxYnJmZHE3c25xMW05ZXc1M28iLCJhY3RpdmVfdXNlcnMiOjQsImV4cCI6MTYyNTY5NTgzMH0.X7mprvTWK0EYrU_29JbhP5fTaZJ2jdf1PwXYxvc8SgU](https://customers.mattermost.com/subscribe/renew?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJsaWNlbnNlX2lkIjoiZTVnY2N3NzFxYnJmZHE3c25xMW05ZXc1M28iLCJhY3RpdmVfdXNlcnMiOjQsImV4cCI6MTYyNTY5NTgzMH0.X7mprvTWK0EYrU_29JbhP5fTaZJ2jdf1PwXYxvc8SgU)
> > Renew error
> > Up for renewal email
> 
> I see the feedback email is missing in the email footer. @marianunez is that due to the fact that it’s blank in the config?

Yes, this was reported in this bug actually https://mattermost.atlassian.net/browse/MM-36603 For this renewal and cloud emails we probably don't want to use the configured one. Left a comment in the ticket for Katie to confirm.

**Comment by stevemudie** [comment]
> @stevemudie let’s give it one more day to make sure those emails don’t come in again and then that will be the final step of the QA

The emails did not arrive again so I'm guessing it's all good.

**Review by stevemudie — APPROVED** [review]
Looks good!

- Created an expiring (61 days to expiration) license
- Uploaded it the test instance (set up on the PR)
- Added an extra admin with an email I could monitor
- Waited for 1 or 2 days and received an email about the license being up for renewal
- Checked again after 24 hours to ensure email was not sent again

**Comment by marianunez** [comment]
@amyblais could we cherry-pick this to the `release-5.37` branch or is it too late to include in the next release?

**Comment by amyblais** [comment]
@marianunez We're already past feature complete for v5.37. If this is urgent, maybe you can share more details about why it would be good to include this in v5.37.

**Comment by marianunez** [comment]
> @marianunez We're already past feature complete for v5.37. If this is urgent, maybe you can share more details about why it would be good to include this in v5.37.

@amyblais No urgency, wasn't sure if feature complete was today with the cloud Fast-Forward but just checked it was the last one. This can wait for the Aug release. Thanks!

**Pull request merged**
